File: ios/Flashmart/ViewModels/FinanceViewModel.swift
import Foundation
import Observation

@Observable
final class FinanceViewModel {
    var loading = false
    var error: String?
    var message: String?
    var period = "month"
    var settlementStatusFilter: String?
    var overview: FinanceOverview?
    var investor: InvestorDashboard?
    var settlements: [DealerSettlement] = []
    var selectedSettlement: DealerSettlement?
    var commissionRules: [CommissionRule] = []
    var auditLogs: [FinanceAuditLog] = []
    var dealerPerformance: DealerPerformance?
    var reportShareURL: URL?

    private let apiClient: MartAPIClient
    private let isoDay: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withFullDate]
        return f
    }()

    init(apiClient: MartAPIClient) {
        self.apiClient = apiClient
    }

    @MainActor
    func setPeriod(_ value: String) {
        period = value
        Task { await refresh() }
    }

    @MainActor
    func setSettlementStatusFilter(_ value: String?) {
        settlementStatusFilter = value
        Task { await refreshSettlements() }
    }

    @MainActor
    func refresh() async {
        loading = true
        error = nil
        do {
            async let overviewTask = apiClient.financeOverview(period: period)
            async let investorTask = apiClient.investorDashboard()
            async let settlementsTask = apiClient.settlements(period: period, status: settlementStatusFilter)
            async let rulesTask = apiClient.commissionRules()
            async let auditTask = apiClient.financeAudit()
            overview = try await overviewTask
            investor = try await investorTask
            settlements = try await settlementsTask
            commissionRules = try await rulesTask
            auditLogs = try await auditTask
            loading = false
        } catch {
            loading = false
            self.error = error.localizedDescription
        }
    }

    @MainActor
    private func refreshSettlements() async {
        do {
            settlements = try await apiClient.settlements(period: period, status: settlementStatusFilter)
        } catch {
            self.error = error.localizedDescription
        }
    }

    @MainActor
    func generateSettlement(dealerId: String) async {
        loading = true
        error = nil
        let end = Date()
        let start = Calendar.current.date(byAdding: .day, value: -30, to: end) ?? end
        do {
            _ = try await apiClient.generateSettlement(
                GenerateSettlementRequest(
                    dealerId: dealerId,
                    startDate: isoDay.string(from: start),
                    endDate: isoDay.string(from: end)
                )
            )
            message = "Settlement generated"
            await refresh()
        } catch {
            loading = false
            self.error = error.localizedDescription
        }
    }

    @MainActor
    func loadSettlement(_ id: String) async {
        do {
            selectedSettlement = try await apiClient.settlementDetail(id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    @MainActor
    func loadDealerPerformance(_ dealerId: String) async {
        loading = true
        error = nil
        dealerPerformance = nil
        do {
            dealerPerformance = try await apiClient.dealerPerformance(dealerId, period: period)
            loading = false
        } catch {
            loading = false
            self.error = error.localizedDescription
        }
    }

    @MainActor
    func recordPayment(
        settlementId: String,
        amount: Double,
        method: String,
        utr: String,
        reference: String,
        remarks: String
    ) async -> Bool {
        loading = true
        error = nil
        do {
            let updated = try await apiClient.recordSettlementPayment(
                settlementId,
                body: RecordSettlementPaymentRequest(
                    amount: amount,
                    paymentMethod: method,
                    utrNumber: utr.isEmpty ? nil : utr,
                    transactionReference: reference.isEmpty ? nil : reference,
                    paymentDate: isoDay.string(from: Date()),
                    remarks: remarks.isEmpty ? nil : remarks
                )
            )
            selectedSettlement = updated
            message = "Payment recorded"
            await refresh()
            return true
        } catch {
            loading = false
            self.error = error.localizedDescription
            return false
        }
    }

    @MainActor
    func updateGlobalCommission(rate: Double) async {
        await upsertRule(
            UpsertCommissionRuleRequest(ruleType: "GLOBAL", rate: rate, dealerId: nil, productId: nil),
            message: "Global commission updated to \(rate)%"
        )
    }

    @MainActor
    func updateDealerCommission(dealerId: String, rate: Double) async {
        await upsertRule(
            UpsertCommissionRuleRequest(ruleType: "DEALER", rate: rate, dealerId: dealerId, productId: nil),
            message: "Dealer commission updated to \(rate)%"
        )
    }

    @MainActor
    func updateProductCommission(productId: String, rate: Double) async {
        await upsertRule(
            UpsertCommissionRuleRequest(ruleType: "PRODUCT", rate: rate, dealerId: nil, productId: productId),
            message: "Product commission updated to \(rate)%"
        )
    }

    @MainActor
    private func upsertRule(_ body: UpsertCommissionRuleRequest, message: String) async {
        do {
            _ = try await apiClient.upsertCommissionRule(body)
            self.message = message
            await refresh()
        } catch {
            self.error = error.localizedDescription
        }
    }

    @MainActor
    func backfillRevenues() async {
        do {
            let res = try await apiClient.backfillRevenues()
            message = "Backfilled \(res.created ?? 0) revenue records"
            await refresh()
        } catch {
            self.error = error.localizedDescription
        }
    }

    @MainActor
    func downloadReport(type: String, format: String) async {
        loading = true
        error = nil
        do {
            reportShareURL = try await apiClient.downloadFinanceReport(type: type, format: format, period: period)
            message = "Report ready to share"
            loading = false
        } catch {
            loading = false
            self.error = error.localizedDescription
        }
    }

    func clearReportShareURL() {
        reportShareURL = nil
    }

    func clearMessage() {
        message = nil
    }
}

File: android/app/src/main/java/com/mart/distribution/demo/feature/finance/FinanceViewModel.kt
package com.mart.distribution.demo.feature.finance

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.mart.distribution.demo.data.api.MartApi
import com.mart.distribution.demo.data.api.dto.CommissionRuleDto
import com.mart.distribution.demo.data.api.dto.DealerPerformanceDto
import com.mart.distribution.demo.data.api.dto.DealerSettlementDto
import com.mart.distribution.demo.data.api.dto.FinanceAuditLogDto
import com.mart.distribution.demo.data.api.dto.FinanceOverviewDto
import com.mart.distribution.demo.data.api.dto.GenerateSettlementRequest
import com.mart.distribution.demo.data.api.dto.InvestorDashboardDto
import com.mart.distribution.demo.data.api.dto.RecordSettlementPaymentRequest
import com.mart.distribution.demo.data.api.dto.UpsertCommissionRuleRequest
import com.mart.distribution.demo.util.ApiErrorMessages
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.LocalDate

data class FinanceReportDownload(
    val bytes: ByteArray,
    val filename: String,
    val mimeType: String,
)

data class FinanceUiState(
    val loading: Boolean = false,
    val error: String? = null,
    val period: String = "month",
    val settlementStatusFilter: String? = null,
    val overview: FinanceOverviewDto? = null,
    val investor: InvestorDashboardDto? = null,
    val settlements: List<DealerSettlementDto> = emptyList(),
    val selectedSettlement: DealerSettlementDto? = null,
    val commissionRules: List<CommissionRuleDto> = emptyList(),
    val auditLogs: List<FinanceAuditLogDto> = emptyList(),
    val dealerPerformance: DealerPerformanceDto? = null,
    val message: String? = null,
    val reportDownload: FinanceReportDownload? = null,
)

class FinanceViewModel(
    private val martApi: MartApi,
) : ViewModel() {
    private val _ui = MutableStateFlow(FinanceUiState())
    val uiState: StateFlow<FinanceUiState> = _ui.asStateFlow()

    fun setPeriod(period: String) {
        _ui.update { it.copy(period = period) }
        refresh()
    }

    fun setSettlementStatusFilter(status: String?) {
        _ui.update { it.copy(settlementStatusFilter = status) }
        refreshSettlements()
    }

    fun refresh() {
        viewModelScope.launch {
            _ui.update { it.copy(loading = true, error = null) }
            try {
                val period = _ui.value.period
                val status = _ui.value.settlementStatusFilter
                val overview = martApi.financeOverview(period = period)
                val investor = martApi.investorDashboard()
                val settlements = martApi.settlements(period = period, status = status)
                val rules = martApi.commissionRules()
                val audit = martApi.financeAudit()
                _ui.update {
                    it.copy(
                        loading = false,
                        overview = overview,
                        investor = investor,
                        settlements = settlements,
                        commissionRules = rules,
                        auditLogs = audit,
                    )
                }
            } catch (e: Exception) {
                _ui.update {
                    it.copy(loading = false, error = ApiErrorMessages.fromThrowable(e, "Could not load finance data"))
                }
            }
        }
    }

    private fun refreshSettlements() {
        viewModelScope.launch {
            try {
                val period = _ui.value.period
                val status = _ui.value.settlementStatusFilter
                val settlements = martApi.settlements(period = period, status = status)
                _ui.update { it.copy(settlements = settlements) }
            } catch (e: Exception) {
                _ui.update { it.copy(error = ApiErrorMessages.fromThrowable(e, "Could not load settlements")) }
            }
        }
    }

    fun generateSettlement(dealerId: String, onDone: () -> Unit = {}) {
        val end = LocalDate.now()
        val start = end.minusDays(30)
        viewModelScope.launch {
            _ui.update { it.copy(loading = true, error = null) }
            try {
                martApi.generateSettlement(
                    GenerateSettlementRequest(
                        dealerId = dealerId,
                        startDate = start.toString(),
                        endDate = end.toString(),
                    ),
                )
                refresh()
                _ui.update { it.copy(message = "Settlement generated") }
                onDone()
            } catch (e: Exception) {
                _ui.update {
                    it.copy(loading = false, error = ApiErrorMessages.fromThrowable(e, "Could not generate settlement"))
                }
            }
        }
    }

    fun loadSettlement(id: String) {
        viewModelScope.launch {
            try {
                val detail = martApi.settlementDetail(id)
                _ui.update { it.copy(selectedSettlement = detail) }
            } catch (e: Exception) {
                _ui.update { it.copy(error = ApiErrorMessages.fromThrowable(e, "Could not load settlement")) }
            }
        }
    }

    fun loadDealerPerformance(dealerId: String) {
        viewModelScope.launch {
            _ui.update { it.copy(loading = true, error = null, dealerPerformance = null) }
            try {
                val perf = martApi.dealerPerformance(dealerId, period = _ui.value.period)
                _ui.update { it.copy(loading = false, dealerPerformance = perf) }
            } catch (e: Exception) {
                _ui.update {
                    it.copy(loading = false, error = ApiErrorMessages.fromThrowable(e, "Could not load dealer performance"))
                }
            }
        }
    }

    fun recordPayment(
        settlementId: String,
        amount: Double,
        method: String,
        utr: String,
        reference: String,
        remarks: String,
        onDone: () -> Unit,
    ) {
        viewModelScope.launch {
            _ui.update { it.copy(loading = true, error = null) }
            try {
                val updated =
                    martApi.recordSettlementPayment(
                        settlementId,
                        RecordSettlementPaymentRequest(
                            amount = amount,
                            paymentMethod = method,
                            utrNumber = utr.ifBlank { null },
                            transactionReference = reference.ifBlank { null },
                            paymentDate = LocalDate.now().toString(),
                            remarks = remarks.ifBlank { null },
                        ),
                    )
                _ui.update { it.copy(selectedSettlement = updated, message = "Payment recorded", loading = false) }
                refresh()
                onDone()
            } catch (e: Exception) {
                _ui.update {
                    it.copy(loading = false, error = ApiErrorMessages.fromThrowable(e, "Could not record payment"))
                }
            }
        }
    }

    fun updateGlobalCommission(rate: Double) {
        upsertRule(UpsertCommissionRuleRequest(ruleType = "GLOBAL", rate = rate), "Global commission updated to $rate%")
    }

    fun updateDealerCommission(dealerId: String, rate: Double) {
        upsertRule(
            UpsertCommissionRuleRequest(ruleType = "DEALER", rate = rate, dealerId = dealerId),
            "Dealer commission updated to $rate%",
        )
    }

    fun updateProductCommission(productId: String, rate: Double) {
        upsertRule(
            UpsertCommissionRuleRequest(ruleType = "PRODUCT", rate = rate, productId = productId),
            "Product commission updated to $rate%",
        )
    }

    private fun upsertRule(body: UpsertCommissionRuleRequest, successMessage: String) {
        viewModelScope.launch {
            try {
                martApi.upsertCommissionRule(body)
                refresh()
                _ui.update { it.copy(message = successMessage) }
            } catch (e: Exception) {
                _ui.update { it.copy(error = ApiErrorMessages.fromThrowable(e, "Could not update commission")) }
            }
        }
    }

    fun backfillRevenues() {
        viewModelScope.launch {
            try {
                val res = martApi.backfillRevenues()
                _ui.update { it.copy(message = "Backfilled ${res["created"] ?: 0} revenue records") }
                refresh()
            } catch (e: Exception) {
                _ui.update { it.copy(error = ApiErrorMessages.fromThrowable(e, "Backfill failed")) }
            }
        }
    }

    fun downloadReport(type: String, format: String) {
        viewModelScope.launch {
            _ui.update { it.copy(loading = true, error = null) }
            try {
                val body = martApi.downloadReport(type = type, format = format, period = _ui.value.period)
                val bytes = body.bytes()
                val ext = if (format == "xlsx") "xlsx" else "csv"
                val mime = if (format == "xlsx") "application/vnd.ms-excel" else "text/csv"
                _ui.update {
                    it.copy(
                        loading = false,
                        reportDownload = FinanceReportDownload(bytes, "$type-report.$ext", mime),
                        message = "Report ready to share",
                    )
                }
            } catch (e: Exception) {
                _ui.update {
                    it.copy(loading = false, error = ApiErrorMessages.fromThrowable(e, "Could not download report"))
                }
            }
        }
    }

    fun clearReportDownload() {
        _ui.update { it.copy(reportDownload = null) }
    }

    fun clearMessage() {
        _ui.update { it.copy(message = null) }
    }
}
